trees/trees_test.go

```diff
@@ -0,0 +1,369 @@
+			wantErr:     true,
+		},
+		{
+			desc:        "frozenTree",
+			treeID:      frozenTree.TreeId,
+			opts:        GetOpts{TreeType: trillian.TreeType_LOG, Readonly: true},
+			storageTree: &frozenTree,
+			wantTree:    &frozenTree,
+		},
+		{
+			desc:        "frozenTreeNotReadonly",
+			treeID:      frozenTree.TreeId,
+			opts:        GetOpts{TreeType: trillian.TreeType_LOG},
+			storageTree: &frozenTree,
+			wantErr:     true,
+		},
+		{
+			desc:        "softDeleted",
+			treeID:      softDeletedTree.TreeId,
+			opts:        GetOpts{TreeType: trillian.TreeType_LOG},
+			storageTree: &softDeletedTree,
+			wantErr:     true,
+		},
+		{
+			desc:        "hardDeleted",
+			treeID:      hardDeletedTree.TreeId,
+			opts:        GetOpts{TreeType: trillian.TreeType_LOG},
+			storageTree: &hardDeletedTree,
+			wantErr:     true,
+		},
+		{
+			desc:     "treeInCtx",
+			treeID:   logTree.TreeId,
+			opts:     GetOpts{TreeType: trillian.TreeType_LOG},
+			ctxTree:  &logTree,
+			wantTree: &logTree,
+		},
+		{
+			desc:        "wrongTreeInCtx",
+			treeID:      logTree.TreeId,
+			opts:        GetOpts{TreeType: trillian.TreeType_LOG},
+			ctxTree:     &mapTree,
+			storageTree: &logTree,
+			wantTree:    &logTree,
+		},
+		{
+			desc:     "beginErr",
+			treeID:   logTree.TreeId,
+			opts:     GetOpts{TreeType: trillian.TreeType_LOG},
+			beginErr: true,
+			wantErr:  true,
+		},
+		{
+			desc:    "getErr",
+			treeID:  logTree.TreeId,
+			opts:    GetOpts{TreeType: trillian.TreeType_LOG},
+			getErr:  true,
+			wantErr: true,
+		},
+		{
+			desc:      "commitErr",
+			treeID:    logTree.TreeId,
+			opts:      GetOpts{TreeType: trillian.TreeType_LOG},
+			commitErr: true,
+			wantErr:   true,
+		},
+	}
+
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	for _, test := range tests {
+		ctx := NewContext(context.Background(), test.ctxTree)
+
+		var beginErr, getErr, commitErr error
+		if test.beginErr {
+			beginErr = errors.New("begin error")
+		}
+		if test.getErr {
+			getErr = errors.New("get error")
+		}
+		if test.commitErr {
+			commitErr = errors.New("commit error")
+		}
+
+		admin := storage.NewMockAdminStorage(ctrl)
+		tx := storage.NewMockReadOnlyAdminTX(ctrl)
+		admin.EXPECT().Snapshot(ctx).MaxTimes(1).Return(tx, beginErr)
+		tx.EXPECT().GetTree(ctx, test.treeID).MaxTimes(1).Return(test.storageTree, getErr)
+		tx.EXPECT().Close().MaxTimes(1).Return(nil)
+		tx.EXPECT().Commit().MaxTimes(1).Return(commitErr)
+
+		tree, err := GetTree(ctx, admin, test.treeID, test.opts)
+		if hasErr := err != nil; hasErr != test.wantErr {
+			t.Errorf("%v: GetTree() = (_, %q), wantErr = %v", test.desc, err, test.wantErr)
+			continue
+		} else if hasErr {
+			continue
+		}
+
+		if !proto.Equal(tree, test.wantTree) {
+			diff := pretty.Compare(tree, test.wantTree)
+			t.Errorf("%v: post-GetTree diff:\n%v", test.desc, diff)
+		}
+	}
+}
+
+func TestHash(t *testing.T) {
+	tests := []struct {
+		hash     sigpb.DigitallySigned_HashAlgorithm
```

Comment:
(nit) rename to `hashAlgo` or somesuch so you don't have two things named `..hash` of different types.